Micro-opt of ArrayDeque.remove

isaacl opened the pull request:
Move a bounds check outside of a remove loop and remove
a redundant set op. These should improve performance,
especially when the index is close to head or tail.

isaacl commented:
@szeiger ^

Ichoran reviewed (commented):
You have to actually benchmark this one to know whether it's faster.  The reason is that avoiding bounds checks and optimizing well are at least as important as avoiding (inexpensive) work.

isaacl commented:
hmm, OK but the change strictly reduces the number of checks. But fair enough.

Ichoran commented:
Hm, actually, the bounds check thing is probably irrelevant given the ring buffer logic.  Let me think about it a moment longer.

isaacl commented:
generally the change is splitting each while loop into two parts, the part that does one type of set and the part that does the other type of set. We're looping through the same total number of values.

Ichoran commented on src/library/scala/collection/mutable/ArrayDeque.scala:
```diff
@@ -248,15 +248,23 @@ class ArrayDeque[A] protected (
         while(i >= 0) {
           i -= 1
-          _set(i, if (i >= removals) _get(i - removals) else null.asInstanceOf[A])
+          _set(i, null.asInstanceOf[A])
```
I don't understand this logic.  Why set -1 to `null.asInstanceOf[A]`?  I guess it probably doesn't hurt since there has to be a gap there (one element isn't used).

Ichoran replied:
(I'm thinking about the case where `i == 0` starting out.)

isaacl replied:
yeah it was confusing to me too.  For simple cases it does one additional null set, but I think it's on purpose to handle nulling out a value that moves

It's undocumented and not well tested -- I had a typo in an earlier version and the tests still passed...

Ichoran replied:
That's what I thought at first, but `i == -1` means _one element before the start_, which was presumably null to begin with, and shouldn't have been touched by the remove operation.  So it looks to me like probably an off-by-one error, though not one that matters because there's a dead element anyway?  Or maybe it was copied from some other context where it _was_ important to write a null element before start?

Ichoran replied:
(i'm also _slighly_ worried by not being confident about what's going on here in light of your other change which reduces the extra space by one cell.  I _think_ it's still fine, and I _also_ think this is just a bug, but my confidence is middling at best because there is so much detailed indexing in this implementation (as is typically necessary for decent performance in a ring buffer).)

isaacl replied:
i == -1 wraps around and sets the null sentinel at the end to null

isaacl replied:
~~Yeah I guess I wasn’t 100% sure but I think it’s on purpose. The intention in this PR is to not change any semantics, so that only a benchmark is needed.~~

isaacl replied:
I've updated the patch to fix this as well.

diesalbla replied:
@isaacl Since off-by-one errors are always worrying, are there any unit tests covering the cases in which an `ArrayDequeue` is full and up-scaled, in which the ends are both in the beginning or the end of the array?

isaacl replied:
not checked in but the benchmark tests that

Ichoran commented:
I'm pretty sure I still can't predict which way is better, even given that the indexing isn't raw (and thus bounds likely get checked anyway.  The reason being that the optimizer often produces better results with simpler logic.

isaacl commented:
fair enough, I'll run a bench

isaacl commented:
@pathikrit what do you think about this patch? Is your benchmark setup handy? (I can run one if not..)

We're also wondering about the [first while loop in remove](https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/mutable/ArrayDeque.scala#L249-L252):

```scala
while(i >= 0) {
  i -= 1
   _set(i, if (i >= removals) _get(i - removals) else null.asInstanceOf[A])
}
```

which will often null out 1 index beyond the removal range.  is that on purpose?

pathikrit commented:
@isaacl : Yup please use the benchmark suite [from here](https://github.com/pathikrit/collection-strawman/blob/0519263c9fe8fb57820b01a76852a0dd97348786/benchmarks/time/src/main/scala/strawman/collection/mutable/ArrayDequeBenchmark.scala) - it has 0 dependencies and gets you close to 100% code coverage on ArrayDequeue

And, yes, you are probably right - it is doing 1 extra nullifying

isaacl committed:
Micro-opt of ArrayDeque.remove

Move a bounds check outside of a remove loop and remove
a redundant set op. These should improve performance,
especially when the index is close to head or tail.

isaacl commented:
I ran the benchmark and see a nice speedup, which held across multiple runs.

```===============[Drop some items from an head index]=================
  ArrayDeque:     0.78 ms
 ArrayDeque2:     0.66 ms
===============[Drop some items from a tail index]=================
  ArrayDeque:     0.76 ms
 ArrayDeque2:     0.39 ms```
```

I have also amended my code to stop setting the null sentinel each remove.

Ichoran reviewed (approved):
Benchmarks look good!

ashawley commented:
I tried to verify this optimization, but couldn't confirm it.

I ran a version of the JMH benchmark in #8211, that removes half of a 10k-sized array at various places in the array (head, near head, near tail, ...).  Here are the results:

```
a90e4389ea (HEAD of this branch)
Benchmark                           (length)  Mode  Cnt  Score   Error  Units
ArrayDequeBenchmark.removeFromHead     10000  avgt   20  6.116 ± 1.663  us/op
ArrayDequeBenchmark.removeFromTail     10000  avgt   20  5.878 ± 0.715  us/op
ArrayDequeBenchmark.removeNearHead     10000  avgt   20  5.790 ± 0.391  us/op
ArrayDequeBenchmark.removeNearTail     10000  avgt   20  5.708 ± 0.672  us/op

a37e40399b (HEAD^ of this branch)
Benchmark                           (length)  Mode  Cnt  Score   Error  Units
ArrayDequeBenchmark.removeFromHead     10000  avgt   20  6.211 ± 0.518  us/op
ArrayDequeBenchmark.removeFromTail     10000  avgt   20  5.747 ± 0.903  us/op
ArrayDequeBenchmark.removeNearHead     10000  avgt   20  5.409 ± 0.170  us/op
ArrayDequeBenchmark.removeNearTail     10000  avgt   20  5.547 ± 0.360  us/op
```

isaacl commented:
JMH is obviously a more robust bench, but I'm surprised that my PR doesn't change perf in a statistically significant manner.  The change removes drops the number of branches instructions by ~50% for large deletes (branch in the while condition, but none in the while body).  Even with branch prediction I'd expect some change.

can you link a gist of your JMH?

ashawley commented:
It's in the PR I linked, #8211.

Ichoran commented:
@ashawley - Thanks for checking that with JMH!  I neglected to check what framework was used and had assumed it was doing something reasonable.  (You need to tell JMH to increase the sample size to tell whether the change has a minor positive effect, though; the error numbers are pretty high.)

@isaacl - I wouldn't believe any of the numbers from what you ran.  It's measuring some sort of messy amalgam of class loading time, single-pass optimizability, garbage collection, and who knows what else.

@pathikrit - Please don't advise anyone to do anything with that "benchmark"!  Benchmarking on the JVM is, sadly, very complicated to get right.  The "this is wrong, don't do this" benchmark in this article is considerably more sophisticated than your approach and still gives wrong answers: https://www.oracle.com/technetwork/articles/java/architect-benchmarking-2266277.html

Everyone who happens to be watching:

If you only need to measure the difference between, say, 100ms and 5s, a single-shot time delta can do the job.  Otherwise, you need to break out tools designed to help tame the complexity of the JVM execution model so you can make comparisons.

Another cautionary tale about microbenchmarking (from Brian Goetz!): https://www.ibm.com/developerworks/java/library/j-jtp02225/

isaacl commented:
I fixed a couple bugs in the JMH benchmark and see speedups in [this bench](https://gist.github.com/isaacl/e062768f7b5036e5ad3791b9662beffb/raw) as well.

**PRE 8091**
```
Benchmark                           (length)  (range)  Mode  Cnt  Score   Error  Units
ArrayDequeBenchmark.removeFromHead     10000     1000  avgt   20  0.186 ± 0.015  us/op
ArrayDequeBenchmark.removeFromTail     10000     1000  avgt   20  0.427 ± 0.013  us/op
ArrayDequeBenchmark.removeNearHead     10000     1000  avgt   20  0.560 ± 0.022  us/op
ArrayDequeBenchmark.removeNearTail     10000     1000  avgt   20  0.517 ± 0.018  us/op
```

**POST 8091**
```
Benchmark                           (length)  (range)  Mode  Cnt  Score   Error  Units
ArrayDequeBenchmark.removeFromHead     10000     1000  avgt   20  0.187 ± 0.016  us/op
ArrayDequeBenchmark.removeFromTail     10000     1000  avgt   20  0.198 ± 0.015  us/op
ArrayDequeBenchmark.removeNearHead     10000     1000  avgt   20  0.291 ± 0.013  us/op
ArrayDequeBenchmark.removeNearTail     10000     1000  avgt   20  0.277 ± 0.017  us/op
```

isaacl commented:
oh and the reason you didn't see anything is that removing **half** of the Array will trigger the reallocation branch of remove, and my code does not affect that branch. So it makes sense your JMH saw nothing...

Ichoran commented:
Looks great!

pathikrit commented:
> @pathikrit - Please don't advise anyone to do anything with that "benchmark"! 

Sorry for the confusion - I recommend _that_ benchmark not for performance checks but to do correctness checks since it pits `ArrayDeque` against `ArrayBuffer` and does bunch of "do this and must get same result" type of checks.

Ichoran commented:
@pathikrit - Okay, yes, it is a decent correctness check, just strangely named!
